src/analysis/locally_convex/basic.lean

```diff
@@ -119,7 +96,48 @@ forall_congr $ λ x, ⟨λ ⟨r, hr, hx⟩, ⟨r, hr.le, λ a ha, hx a ha.le⟩,
   ⟨r + 1, add_pos_of_nonneg_of_pos hr zero_lt_one,
     λ a ha, hx ((lt_add_of_pos_right r zero_lt_one).trans_le ha)⟩⟩
 
-end add_comm_group
+variables (𝕜)
+
+/-- A set `A` is balanced if `a • A` is contained in `A` whenever `a` has norm at most `1`. -/
+def balanced (A : set E) := ∀ a : 𝕜, ∥a∥ ≤ 1 → a • A ⊆ A
+
+variables {𝕜}
+
+lemma balanced_mem {s : set E} (hs : balanced 𝕜 s) {x : E} (hx : x ∈ s) {a : 𝕜} (ha : ∥a∥ ≤ 1) :
+  a • x ∈ s :=
+mem_of_subset_of_mem (hs a ha) (smul_mem_smul_set hx)
+
+lemma balanced_univ : balanced 𝕜 (univ : set E) := λ a ha, subset_univ _
+
+lemma balanced.union (hA : balanced 𝕜 A) (hB : balanced 𝕜 B) : balanced 𝕜 (A ∪ B) :=
+begin
+  intros a ha t ht,
+  rw smul_set_union at ht,
+  exact ht.imp (λ x, hA _ ha x) (λ x, hB _ ha x),
+end
+
+lemma balanced.inter (hA : balanced 𝕜 A) (hB : balanced 𝕜 B) : balanced 𝕜 (A ∩ B) :=
+begin
+  rintro a ha _ ⟨x, ⟨hx₁, hx₂⟩, rfl⟩,
+  exact ⟨hA _ ha ⟨_, hx₁, rfl⟩, hB _ ha ⟨_, hx₂, rfl⟩⟩,
+end
+
+end has_scalar
+
+section add_comm_monoid
+variables [add_comm_monoid E] [module 𝕜 E] {s t u v A B : set E}
```

Comment:
You can deduce `add_comm_group E` from your hypotheses (setting `-x = -1 • x`), so we decided a while ago that we wouldn't bother "generalizing" in this direction. cc @ocfnash

Reply:
that's true, I was getting carried away :smile: